src/Storages/StorageFile.cpp

```diff
@@ -577,29 +716,102 @@ ColumnsDescription StorageFile::getTableStructureFromFile(
             "Cannot extract table structure from {} format file, because there are no files with provided path. "
             "You must specify table structure manually", format);
 
-    std::optional<ColumnsDescription> columns_from_cache;
-    if (context->getSettingsRef().schema_inference_use_cache_for_file)
-        columns_from_cache = tryGetColumnsFromCache(paths, format, format_settings, context);
-
     ColumnsDescription columns;
-    if (columns_from_cache)
+    if (archive_info)
     {
-        columns = *columns_from_cache;
+        std::vector<std::string> paths_for_schema_cache;
+        std::optional<ColumnsDescription> columns_from_cache;
+
+        if (context->getSettingsRef().schema_inference_use_cache_for_file)
+        {
+            paths_for_schema_cache.reserve(archive_info->paths_to_archives.size());
+            struct stat file_stat{};
+            for (const auto & archive : archive_info->paths_to_archives)
+            {
+                const auto & full_path = paths_for_schema_cache.emplace_back(fmt::format("{}::{}", archive, archive_info->path_in_archive));
+
+                if (!columns_from_cache)
+                {
+                    auto & schema_cache = getSchemaCache(context);
+                    auto get_last_mod_time = [&]() -> std::optional<time_t>
+                    {
+                        if (0 != stat(archive.c_str(), &file_stat))
+                            return std::nullopt;
+
+                        return file_stat.st_mtime;
+                    };
+
+                    auto cache_key = getKeyForSchemaCache(full_path, format, format_settings, context);
+                    columns_from_cache = schema_cache.tryGetColumns(cache_key, get_last_mod_time);
+                }
+            }
+        }
+
+        if (columns_from_cache)
+        {
+            columns = std::move(*columns_from_cache);
+        }
+        else
+        {
+            ReadBufferFromArchiveIterator read_buffer_iterator(*archive_info, format, format_settings, context);
+            try
+            {
+                columns = readSchemaFromFormat(
+                    format,
+                    format_settings,
+                    read_buffer_iterator,
+                    /*retry=*/archive_info->paths_to_archives.size() > 1 || !archive_info->isSingleFileRead(),
+                    context);
+            }
+            catch (const DB::Exception & e)
+            {
+                /// maybe we found something in cache while iterating files
+                if (e.code() == ErrorCodes::CANNOT_EXTRACT_TABLE_STRUCTURE)
+                {
+                    if (read_buffer_iterator.columns_from_cache)
+                        columns = std::move(*read_buffer_iterator.columns_from_cache);
+                    else
+                        throw;
+                }
+                else
+                {
+                    throw;
+                }
+            }
```

Comment:
No need to do it actually, `IReadBufferIterator` has virtual method `getCachedColumns` that is used in `readSchemaFromFormat`:
https://github.com/ClickHouse/ClickHouse/blob/fc7a8d0080522b4f92a43be0848b44c5478a193d/src/Formats/ReadSchemaUtils.cpp#L182-L183

So readSchemaFromFormat will return cached columns and no try/catch is needed